// mixed-map/bld/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MixedMap = /** @class */ (function () {
    function MixedMap() {
    }
    MixedMap.prototype.has = function (key) {
        return this.getMap(key).has(key);
    };
    MixedMap.prototype.get = function (key) {
        return this.getMap(key).get(key);
    };
    MixedMap.prototype.set = function (key, value) {
        this.getMap(key).set(key, value);
        return this;
    };
    MixedMap.prototype.delete = function (key) {
        return this.getMap(key).delete(key);
    };
    MixedMap.prototype.getMap = function (key) {
        if (typeof key === 'object' && key !== null) {
            if (this.weakMap) {
                return this.weakMap;
            }
            else {
                return this.weakMap = new WeakMap();
            }
        }
        else {
            if (this.map) {
                return this.map;
            }
            else {
                return this.map = new Map();
            }
        }
    };
    return MixedMap;
}());
exports.MixedMap = MixedMap;
exports.default = MixedMap;
